azure/meta.go

```diff
@@ -0,0 +1,131 @@
+package azure
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hairyhenderson/gomplate/v4/env"
+)
+
+// DefaultEndpoint is the IMDS endpoint at Azure
+// https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service
+var DefaultEndpoint = "http://169.254.169.254"
+
+var (
+	// co is a ClientOptions populated from the environment.
+	co ClientOptions
+	// coInit ensures that `co` is only set once.
+	coInit sync.Once
+)
+
+// ClientOptions contains various user-specifiable options for a MetaClient.
+type ClientOptions struct {
+	Timeout time.Duration
+}
+
+// GetClientOptions - Centralised reading of AZURE_TIMEOUT
+// ... but cannot use in vault/auth.go as different strconv.Atoi error handling
+func GetClientOptions() ClientOptions {
+	coInit.Do(func() {
+		timeout := env.Getenv("AZURE_TIMEOUT")
+		if timeout == "" {
+			timeout = "500"
+		}
+
+		t, err := strconv.Atoi(timeout)
+		if err != nil {
+			panic(fmt.Errorf("invalid AZURE_TIMEOUT value '%s' - must be an integer: %w", timeout, err))
+		}
+
+		co.Timeout = time.Duration(t) * time.Millisecond
+	})
+	return co
+}
+
+// MetaClient is used to access metadata accessible via the Azure compute instance
+// metadata service version 1.
+type MetaClient struct {
+	client   *http.Client
+	cache    map[string]string
+	endpoint string
+	options  ClientOptions
+}
+
+// NewMetaClient constructs a new MetaClient with the given ClientOptions. If the environment
+// contains a variable named `AZURE_META_ENDPOINT`, the client will address that, if not the
+// value of `DefaultEndpoint` is used.
+func NewMetaClient(options ClientOptions) *MetaClient {
+	endpoint := env.Getenv("AZURE_META_ENDPOINT")
+	if endpoint == "" {
+		endpoint = DefaultEndpoint
+	}
+
+	return &MetaClient{
+		cache:    make(map[string]string),
+		endpoint: endpoint,
+		options:  options,
+	}
+}
+
+// Meta retrieves a value from the Azure Instance Metadata Service, returning the given default
+// if the service is unavailable or the requested URL does not exist.
+func (c *MetaClient) Meta(key, format, apiVersion string, def ...string) (string, error) {
+	url := c.endpoint + "/metadata/instance/" + key + "?api-version=" + apiVersion + "&format=" + format
+	return c.retrieveMetadata(url, def...)
+}
+
+// retrieveMetadata executes an HTTP request to the Azure Instance Metadata Service with the
+// correct headers set, and extracts the returned value.
+func (c *MetaClient) retrieveMetadata(url string, def ...string) (string, error) {
+	if value, ok := c.cache[url]; ok {
+		return value, nil
+	}
+
+	if c.client == nil {
+		timeout := c.options.Timeout
+		if timeout == 0 {
+			timeout = 500 * time.Millisecond
+		}
+		c.client = &http.Client{Timeout: timeout, Transport: &http.Transport{Proxy: nil}}
+	}
+
+	request, err := http.NewRequest(http.MethodGet, url, nil)
```

Comment:
once `ctx` is passed into `retrieveMetadata`:

```suggestion
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
```